txnsync/peer.go

```diff
@@ -497,7 +506,76 @@ func incomingPeersOnly(peers []*Peer) (incomingPeers []*Peer) {
 
 // incoming related functions
 
-func (p *Peer) addIncomingBloomFilter(round basics.Round, incomingFilter *testableBloomFilter, currentRound basics.Round) {
+// addIncomingBloomFilter keeps the most recent {maxIncomingBloomFilterHistory} filters
+func (p *Peer) addIncomingBloomFilter(round basics.Round, incomingFilter *testableBloomFilter, currentRound basics.Round, maxIncomingFilterElements int) {
+	minRound := currentRound.SubSaturate(2)
+	if round < minRound {
+		// ignore data from the past
+		return
+	}
+	bf := incomingBloomFilter{
+		round:  round,
+		filter: incomingFilter,
+	}
+	elemOk := func(i int) bool {
+		ribf := p.recentIncomingBloomFilters[i]
+		if ribf.filter == nil {
+			return false
+		}
+		if ribf.round < minRound {
+			return false
+		}
+		if incomingFilter.clearPrevious && ribf.filter.encodingParams.Offset == incomingFilter.encodingParams.Offset && ribf.filter.encodingParams.Modulator == incomingFilter.encodingParams.Modulator {
+			return false
+		}
+		return true
+	}
+	// compact the prior list to the front of the array.
+	// order doesn't matter.
+	pos := 0
+	last := len(p.recentIncomingBloomFilters) - 1
+	oldestRound := currentRound
+	for pos <= last {
+		if elemOk(pos) {
+			if p.recentIncomingBloomFilters[pos].round < oldestRound {
+				oldestRound = p.recentIncomingBloomFilters[pos].round
+			}
+			pos++
+			continue
+		}
+		p.recentIncomingBloomFilters[pos] = p.recentIncomingBloomFilters[last]
+		p.recentIncomingBloomFilters[last].filter = nil // GC
+		last--
+	}
+	p.recentIncomingBloomFilters = p.recentIncomingBloomFilters[:last+1]
+	// Simple case: append
+	if last+1 < maxIncomingBloomFilterHistory {
+		p.recentIncomingBloomFilters = append(p.recentIncomingBloomFilters, bf)
+		return
+	}
+	// Too much traffic case: random eviction from oldest round held
+	numAtOldestRound := 0
+	for _, ribf := range p.recentIncomingBloomFilters {
+		if ribf.round == oldestRound {
+			numAtOldestRound++
+		}
+	}
+	// fast simple rand is fine, don't need crypto rand
+	evictIndex := rand.Intn(numAtOldestRound)
+	ari := 0
+	for i, ribf := range p.recentIncomingBloomFilters {
+		if ribf.round == oldestRound {
+			if ari == evictIndex {
+				p.recentIncomingBloomFilters[i] = bf
+				return
+			}
+			ari++
+		}
+	}
+	panic("p.recentIncomingBloomFilters changed out from under addIncomingBloomFilter")
```

Comment:
Please try to avoid using randomness here - it would cause the emulator to generate random results, which would make it completely useless. Instead, try to use some constant logic.
More specifically, I think that we want to evict first `p.recentIncomingBloomFilters` that share the same encoding parameters, and then other encoding parameters for the oldestRound.